waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/ParallelActionScheduler.scala

```diff
@@ -64,33 +65,34 @@ class ParallelActionScheduler(val pools: Map[String, ExecutionPoolDesc]
 
   override def hasRunningActions: Boolean = pools.exists(kv => kv._2.running.nonEmpty)
 
-  override def waitToFinish(flowContext: FlowContext, flowReporter: FlowReporter): Try[(ActionScheduler, Seq[(DataFlowAction, Try[ActionResult])])] = {
-    Try {
-      var finished: Option[Seq[futureResult]] = None
-      do {
-        val runningActionsCount = pools.map(_._2.running.size).sum
-        logInfo(s"Waiting for an action to finish to continue. Running actions: $runningActionsCount")
-        finished = Option(actionFinishedNotificationQueue.poll(1, TimeUnit.MINUTES))
-          .map { oneAction =>
-            //optimistic attempt to get results for more actions without blocking
-            val bucket = new util.LinkedList[futureResult]()
-            actionFinishedNotificationQueue.drainTo(bucket, 1000)
-            bucket.asScala :+ oneAction
+
+  override def waitToFinish(flowContext: FlowContext, flowReporter: FlowReporter): (ActionScheduler, Seq[(DataFlowAction, Try[ActionResult])]) = {
+    @tailrec
+    def loop(finished: Option[Seq[futureResult]]): (ActionScheduler, Seq[(DataFlowAction, Try[ActionResult])]) = {
```

Comment:
Could you case the type properly i.e. `FutureResult`? I know you didn't introduce it...